tracer/build/_build/Build.GitHub.cs

```diff
@@ -346,60 +412,93 @@ partial class Build
 
-            var branch = $"refs/tags/v{FullVersion}";
+            // Get all the builds to TargetBranch that were triggered by a CI push
+
+            var builds = await buildHttpClient.GetBuildsAsync(
+                             project: AzureDevopsProjectId,
+                             definitions: new[] { AzureDevopsConsolidatePipelineId },
+                             reasonFilter: BuildReason.IndividualCI,
+                             branchName: TargetBranch,
+                             queryOrder: BuildQueryOrder.QueueTimeDescending);
+
+            if (builds.Count == 0)
+            {
+                Logger.Error($"::error::No builds found for {TargetBranch}. Did you include the full git ref, e.g. refs/heads/master?");
+                throw new Exception($"No builds found for {TargetBranch}");
+            }
+
+            BuildArtifact artifact = null;
+            var artifactName = $"{FullVersion}-release-artifacts";
+
+            Logger.Info($"Checking builds for artifact called: {artifactName}");
+
+            // start from the current commit, and keep looking backwards until we find a commit that has a build
+            // that has successful artifacts. Should only be called from branches with a linear history (i.e. single parent)
+            // This solves a potential issue where we previously selecting a build by start order, not by the actual
+            // git commit order. Generally that shouldn't be an issue, but if we manually trigger builds on master
+            // (which we sometimes do e.g. trying to bisect and issue, or retrying flaky test for coverage reasons),
+            // then we could end up selecting the wrong build.
+            const int maxCommitsBack = 20;
+            for (var i = 0; i < maxCommitsBack; i++)
+            {
+                var commitSha = GitTasks.Git($"log {TargetBranch}~{i} -1 --pretty=%H")
+                                        .FirstOrDefault(x => x.Type == OutputType.Std)
+                                        .Text;
+
+                Logger.Info($"Looking for builds for {commitSha}");
+
+                foreach (var build in builds)
+                {
+                    if (string.Equals(build.SourceVersion, commitSha, StringComparison.OrdinalIgnoreCase))
+                    {
+                        try
+                        {
+                            artifact = await buildHttpClient.GetArtifactAsync(
+                                           project: AzureDevopsProjectId,
+                                           buildId: build.Id,
+                                           artifactName: artifactName);
+
+                            break;
+                        }
+                        catch (ArtifactNotFoundException)
+                        {
+                            Logger.Info($"Could not find {artifactName} artifact for build {build.Id}. Skipping");
```

Comment:
Here, at the first iteration, we are sure we have the last commit that has run on the consolidated pipeline. So if we don't have artifacts, we should break with an error, shouldn't we? 
Indeed, if we don't have artifacts, It means that - for any reason - the pipeline has failed and no artifacts have been generated. So we should fix the pipeline to make sure there are no real issues and relaunch the release afterwards, no?

Reply:
Yes.... you're right, I think we should fail immediately here:thinking_face:

It's ok for some of the other scenarios we have, e.g. fetching latest code coverage. But in this case, the latest build _triggered_ for master should definitely have artifacts I think...
We just can't guarantee that will be the latest commit (as that might have a noop build) so still need to do the "walk back through the parents" bit.

Fixed in https://github.com/DataDog/dd-trace-dotnet/pull/2324/commits/8f9088a64598ad1d68af598f9f4aa399bec85a19